### PR TITLE
[MRG] USLEEP model

### PR DESCRIPTION
**Pull request by Tgnassou**
Add USleep model for sleep staging
Joint work with @l-omar-chehab

### CODE REVIEW

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:11]:
```suggestion
def _crop_tensors_to_match(x1, x2, axis=-1):
```
make this private

File: braindecode/datasets/base.py
Comment (agramfort) [inline braindecode/datasets/base.py:229]:
```suggestion
        super().__init__(list_of_ds)

```

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:26]:
```suggestion
class _EncoderBlock(nn.Module):
```

make it private. Just keep USleep public

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:29]:
```suggestion
                 in_chans=2,
```

@robintibor @hubertjb we need some consistency here

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:58]:
```suggestion
class _DecoderBlock(nn.Module):
```

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:106]:
```suggestion
    """Sleep staging architecture from Perslev et al 2021.

    U-Net (autoencoder with skip connections) feature-extractor for sleep staging described in [1]_.

```

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:117]: we need to fix this naming

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:218]: cleanup

File: docs/whats_new.rst
Comment (agramfort) [inline docs/whats_new.rst:58]:
```suggestion
- Adding the USleep model :class:`braindecode.models.USleep` (:gh:`282` by `Theo Gnassounou`_ and `Omar Chehab`_)
```

File: examples/plot_sleep_staging.py
Comment (agramfort) [inline examples/plot_sleep_staging.py:104]: you should revert changes in this file.

File: examples/plot_sleep_staging_sequences.py
Comment (agramfort) [inline examples/plot_sleep_staging_sequences.py:3]: this file should go

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:8]: you should add ref to usleep paper not Chambon

File: examples/plot_sleep_staging.py
Comment (l-omar-chehab) [reply examples/plot_sleep_staging.py:104]: I'm not sure I understand what this means

Comment (agramfort) [reply examples/plot_sleep_staging.py:104]: you should not have touched this file in this PR.

Comment (l-omar-chehab) [reply examples/plot_sleep_staging.py]: absolutely - I just made sure this script is up-to-date with braindecode master so we don't inadvertently change it.

File: braindecode/models/usleep.py
Comment (l-omar-chehab) [reply braindecode/models/usleep.py:117]: Yes. We just changed it to in_chans for consistency with (some of the other) braindecode models. We will open a more general PR for naming.

Comment (agramfort) [reply braindecode/models/usleep.py:117]: @l-omar-chehab in_chans now

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:127]:
```suggestion
        Set to 0.070 s in [1]_ (9 samples at sfreq=128).
```

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:143]:
```suggestion
        If True, apply batch normalization after temporal convolutional layers.
```

File: braindecode/models/usleep.py
Comment (agramfort) [inline braindecode/models/usleep.py:147]: put full author list

File: examples/plot_sleep_staging.py
Comment (agramfort) [inline examples/plot_sleep_staging.py:4]: just copy the file 

https://raw.githubusercontent.com/braindecode/braindecode/master/examples/plot_sleep_staging.py

to avoid the diff here

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:12]: put full author list

File: braindecode/models/usleep.py
Comment (l-omar-chehab) [reply braindecode/models/usleep.py:29]: Ok. I changed n_channels to in_chans in the final model (USleep) to homogenize with other braindecode models. 

For the intermediate classes (_EncoderBlock, _DecoderBlock), maybe for now we can leave in_channels and out_channels as is, given they are the names of the attributes in pytorch? And we could address the more general naming issue in a separate PR?

**Comment by agramfort** [comment]
>
> Ok. I changed n_channels to in_chans in the final model (USleep) to
> homogenize with other braindecode models.
>
> For the intermediate classes (_EncoderBlock, _DecoderBlock), maybe for now
> we can leave in_channels and out_channels as is, given they are the names
> of the attributes in pytorch? And we could address the more general naming
> issue in a separate PR?
>

OK

File: examples/plot_sleep_staging.py
Comment (l-omar-chehab) [reply examples/plot_sleep_staging.py]: ok done

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:118]:
```suggestion
from braindecode.preprocessing import zscore
```

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:26]: fix authors

**Comment by agramfort** [comment]
also you need to merge master to fix conflicts in what's new page

**Comment by agramfort** [comment]
looks good https://1291-232335424-gh.circle-artifacts.com/0/dev/auto_examples/plot_sleep_staging_usleep.html !

+1 for MRG on my side

**Review by agramfort — APPROVED** [review]
@hubertjb @gemeinl @robintibor merge if happy

**Comment by hubertjbanville** [comment]
> @hubertjb @gemeinl @robintibor merge if happy

Taking a look now!

File: braindecode/models/usleep.py
Comment (robintibor) [inline braindecode/models/usleep.py:54]: Why not just instantiate these during object construction like
```python
self.pad = nn.ConstantPad1d(padding=1, value=0)
self.pool = nn.MaxPool1d(kernel_size=self.downsample, stride=self.downsample)
```

Seems strange to me to create object to then just use it, then one could also call torch functional API?

**Comment by robintibor** [comment]
I left one comment regarding classes.
Another thing, can we further reduce example doc building time by using less data or fewer epochs? Right now it definitely increases doc build time quite a bit more than all other examples:
image

**Comment by hubertjb** [comment]
I made a few changes, mainly uniformizing the two sleep staging examples. I've changed the preprocessing to fit with the U-Sleep paper, and also tried to limit running time - we'll see if that has the expected effect on the CI.

File: braindecode/preprocessing/preprocess.py
Comment (agramfort) [inline braindecode/preprocessing/preprocess.py:278]: @hubertjbanville I don't get why we have this robust_scale function and changes in plot_sleep_staging.py in this PR. It's not already in master? it's required here for usleep training?

Comment (hubertjb) [reply braindecode/preprocessing/preprocess.py:278]: There wasn't a robust scaling implementation yet in braindecode. It's required here as it's the only preprocessing step used in the U-Sleep paper. We could use a simple zscore but it was simple enough to implement so I added it in.

The changes to `plot_sleep_staging.py` are just because so the two sleep staging examples are more uniform.

**Comment by agramfort** [comment]
@hubertjbanville we will squash and merge this one. We'll have robust scaler and Usleep in one commit. For me it's a bad example for the younger folks here :) can you make a dedicated PR to show good practices? 🙏

**Comment by hubertjb** [comment]
> @hubertjbanville we will squash and merge this one. We'll have robust scaler and Usleep in one commit. For me it's a bad example for the younger folks here :) can you make a dedicated PR to show good practices? pray

Aah right if we squash that would be a mess. Please see #291 ! :)

File: examples/plot_sleep_staging_usleep.py
Comment (robintibor) [inline examples/plot_sleep_staging_usleep.py:71]: so here could be possible to just use scale from sklearn as per https://github.com/braindecode/braindecode/pull/291#issuecomment-872072222 ? or would that be different?

Comment (robintibor) [reply examples/plot_sleep_staging_usleep.py:71]: but could also handle this in separate PR I guess? Then I would merge? Unless you want to change this here already

Comment (gemeinl) [reply examples/plot_sleep_staging_usleep.py:71]: As the comment above suggests, it should be `sklearn.preprocessing.robust_scale`. But yes, it should be possible to use this here instead.

Comment (robintibor) [reply examples/plot_sleep_staging_usleep.py:71]: `sklearn.preprocessing.robust_scale` or `sklearn.preprocessing.scale`?

**Comment by agramfort** [comment]
don't merge yet

>

**Comment by robintibor** [comment]
ok then please reindicate in title when ready to merge

**Comment by hubertjb** [comment]
I also removed the window-wise zscoring step, which isn't in the USleep paper.

Re-reading the paper there's actually one more preprocessing step that is not currently in the example:
> We then clip any value which has an absolute deviation from the median of more than 20 times the IQR of that channel.
Not sure this is necessary for this example but we could add it too.

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:7]: do not do it like this. All sphinx gallery examples have an anchor you can use for a link.
Here this would point to stable release although it should point to dev version.

Anchor is called sphx_glr_auto_examples_plot_sleep_staging.py so I think it could be:
```suggestion
    :ref:`here <sphx_glr_auto_examples_plot_sleep_staging.py>`
```

**Comment by robintibor** [comment]
> I left one comment regarding object creation.
> Another thing, can we further reduce example doc building time by using less data or fewer epochs? Right now it definitely increases doc build time quite a bit more than all other examples:
> image

ah and please also consider about this

File: braindecode/models/usleep.py
Comment (hubertjb) [inline braindecode/models/usleep.py:161]: @l-omar-chehab Is this the number of blocks? Isn't it 12 in the paper?

File: braindecode/models/usleep.py
Comment (hubertjb) [inline braindecode/models/usleep.py:164]: Could this be a parameter? I'm looking for ways to decrease the complexity of the model so the example runs faster...

File: braindecode/models/usleep.py
Comment (hubertjb) [inline braindecode/models/usleep.py:171]: Should there be both `complexity_factor` and `np.sqrt(2)` here?

File: braindecode/models/usleep.py
Comment (Tgnassou) [reply braindecode/models/usleep.py:161]: It's 12 for Usleep because the frequency is 128Hz. With 100Hz it is only possible to take 10 depth because we can divide tensor of size 1

File: braindecode/models/usleep.py
Comment (Tgnassou) [reply braindecode/models/usleep.py:164]: We can reduce the depth yes, if it is your question

File: braindecode/models/usleep.py
Comment (hubertjb) [reply braindecode/models/usleep.py:161]: Ah got it! Maybe this should be explained in the docstring in case someone wants to use the architecture with sfreq=128?

File: braindecode/models/usleep.py
Comment (Tgnassou) [reply braindecode/models/usleep.py:171]: Yes, we just make the same things as USleep

File: braindecode/models/usleep.py
Comment (hubertjb) [reply braindecode/models/usleep.py:164]: OK good, in this case I think it should be added to the arguments!

File: braindecode/models/usleep.py
Comment (hubertjb) [reply braindecode/models/usleep.py:171]: I'm confused by the description in the paper:
image

Each layer has srqt(2) times more convolutional channels than the previous one. For the encoder at least, there's only one conv layer per block. How can the number of degrees of freedom **double** from one block to the next?

Comment (Tgnassou) [reply braindecode/models/usleep.py:171]: Why do you want the degree of freedom double? it is just multiply by sqrt(2)
Maybe i don't understand your point

Comment (hubertjb) [reply braindecode/models/usleep.py:171]: The doubling is just what they say in the paper. It doesn't seem to fit with their equation so that's what I was wondering about. You've looked at their keras implementation - maybe it's clearer in there?

With the current code though, you end up with:
```
np.ceil(5 * np.sqrt(2) * np.sqrt(2) ** np.arange(0, 10 + 1))
Out[5]: array([  8.,  11.,  15.,  21.,  29.,  41.,  57.,  81., 114., 161., 227.])
```
That doesn't look right to me. I think it should be `np.floor` instead of `np.ceil`, and the hardcoded `np.sqrt(2)` should be removed since that's what `complexity_factor` does.

That would give: `np.floor(n_time_filters * complexity_factor ** np.arange(0, depth + 1))`
```
np.floor(5 * np.sqrt(2) ** np.arange(0, 10 + 1))
Out[6]: array([  5.,   7.,  10.,  14.,  20.,  28.,  40.,  56.,  80., 113., 160.])
```

Comment (Tgnassou) [reply braindecode/models/usleep.py:171]: Ok, it's because Omar doesn't like the way USleep is code and he tries somethings different. So it is not exactly the code in keras, i don't know what is the best

Comment (agramfort) [reply braindecode/models/usleep.py:176]: ????? !!!! you mean this architecture is not the published architecture ! and this is done silently as we say in the doc it's USleep???

**Comment by robintibor** [comment]
btw, I would in general advocate to mostly try to reduce running time in examples by reducing number of datapoints and reducing number of epochs (drastically if needed, with clear note this is to reduce running time). This can be done for any example and makes things more uniform across examples and also seems easy to understand to me. Feel this is better than making model smaller which will mean something different for any architecture etc.

**Comment by agramfort** [comment]
+1

>

**Comment by Tgnassou** [comment]
@robintibor 2min46 for the USleep examples, is it ok for you ?

**Comment by robintibor** [comment]
2 min 46 s is fine to me.

However, would prefer if:
* example clearly states whenever a choice is done to make stuff run faster **at that point**, not just at end of example
* as much as possible, as written before, just n_epochs is reduced and data size is reduced, e.g., could you keep model same and just reduce `crop_wake_mins=30` quite a lot?

**Comment by hubertjb** [comment]
I just had a chat with @Tgnassou and @l-omar-chehab. There are three things that still need to be done in this PR:

1. Clarify what hyparameters were actually used in the USleep paper. The paper, the Supplementary Materials and the keras implementation all have different descriptions of how the numbers of convolutional channel are computed. A quick summary:

    * [Paper (p.8)](https://www.nature.com/articles/s41746-021-00440-5): c_t = floor(c_{t-1} * np.sqrt(2)); c_1 = 5
    * [Suppl. (p.32)](https://static-content.springer.com/esm/art%3A10.1038%2Fs41746-021-00440-5/MediaObjects/41746_2021_440_MOESM2_ESM.pdf): c_t = floor(c_{t-1} * np.sqrt(1.67)); c_1 = 6 (doesn't seem to fit with the values in the table)
    * [Keras](https://github.com/perslev/U-Time/blob/usleep-paper-version/utime/models/utime.py#L154) : c_t = floor(h_{t-1} * np.sqrt(2)) where h_t = floor(h_{t-1} * 2); h_1 = ?

@Tgnassou and @l-omar-chehab will reach out to the authors to ask them which version they ended up using.

2. Make sure the model can be used with more than just one fixed input sampling rate. Right now the kernel size and maxpool size are hardcoded to avoid previously seen problems with the dimensions of layers. It looks like the actual problem is that the maxpool size and the depth size (i.e. number of blocks) need to be compatible. @Tgnassou is looking at that.

3. We need another test that compares the dimensions of the model to those reported in the paper.

**Comment by hubertjb** [comment]
Also, to answer your previous comment @robintibor : `crop_wake_mins` actually shouldn't be decreased below 30 (it's the typical value used in the literature). I think at this point the data is really as small as it can get, unless we want to only use a few examples from the two sleep recordings that are used in the example.

**Comment by robintibor** [comment]
Maybe a clarification regarding examples as I see it @hubertjb 
Obviously the nicest thing would be to have the best-performing pipeline as the examples, so people can directly use the code and expect great results and at the same time it is validated to run by the CI. 
However, this would take too long on the CI, and in the end we decided that:
* all decoding examples will run in CI
* code should be as close as possible to good pipeline so users may not be confused and run suboptimal hyperparameters expecting good results
* to ensure fast runs on CI, we drastically reduce data and epochs, even if it makes the example nonsensical as it is and clearly note that these are not useful values, indicating what are the useful values

This way, we can make sure the code always runs, is fast enough on CI, and should prevent users from mistakenly adopting suboptimal pipelines. And it is acceptable if the example due to reduced data and reduced number of examples does not learn anything as is (e.g., our trialwise decoding example (https://braindecode.org/master/auto_examples/plot_bcic_iv_2a_moabb_trial.html)  is barely above chance on train and exactly at chance on valid at the end of training).

Otherwise, with examples that are "more realistic" but with suboptimal hyperparameters, we have seen users adopt suboptimal hyperparameters in their training and expect good results.

**Comment by Tgnassou** [comment]
This PR is ready to merge

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:212]: you need to use nn.CrossEntropyLoss here as you use class-weights? if so can you add a comment
so it's clear why you need to change a default parameter in a first example?

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:17]:
```suggestion
    The number of epochs is here too small and very few recordings are used.
```

File: examples/plot_sleep_staging_usleep.py
Comment (agramfort) [inline examples/plot_sleep_staging_usleep.py:18]: Given what you observe can you be more specific? How many epochs will lead to
what accuracy (roughly). Something like : Taking for example at least 30 recordings
to training and doing 40 epochs should lead to a balanced accuracy close to 0.8.

Comment (agramfort) [reply examples/plot_sleep_staging_usleep.py:18]: can you also put the same warning in plot_sleep_staging.py  ?

File: examples/plot_sleep_staging_usleep.py
Comment (robintibor) [reply examples/plot_sleep_staging_usleep.py:212]: what about these questions @Tgnassou ? please do add some comment about the `depth=12` line before the model creation, so it shows up clearly in the example doc why we reduce here. I am also confused about `apply_softmax=True` and using `nn.CrossEntropyLoss` as doc of cross entropy loss states:
https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html

```
This criterion combines LogSoftmax and NLLLoss in one single class.

[...]

The input is expected to contain raw, unnormalized scores for each class.```

File: examples/plot_sleep_staging_usleep.py
Comment (robintibor) [reply examples/plot_sleep_staging_usleep.py:18]: Yes and also add the changing of the depth parameter, in case that was for execution time @Tgnassou

**Comment by robintibor** [comment]
See comments by me and alex, and also @Tgnassou I invited you to be maintainer, then you don't have to always wait for me to approve workflows. could you accept this invitation?

**Comment by robintibor** [comment]
So I think the loss comment from me above https://github.com/braindecode/braindecode/pull/282#discussion_r672987582 definitely needs addressing. Is somebody still active on it or do we need to search somebody @hubertjb  @Tgnassou @agramfort

**Comment by Tgnassou** [comment]
I am in holidays, that's why i'm not very active these days. But I will continue the work in september! Is it ok for you @robintibor ?

**Comment by robintibor** [comment]
ok my plan was to have a release ~end of august, but maybe we can shift to midseptember anyways, then this one could be inside as well

**Comment by hubertjb** [comment]
I'm still working on it and should have results soon @robintibor

**Comment by hubertjb** [comment]
Yes, I'll share that asap! 

So far I can say that the balanced sequence sampler does seem to speed up learning a bit. The USleep model, especially in its depth=12 version, is massively overparametrized though, especially given a smallish dataset like Sleep Physionet. Reducing the depth and adding some weight decay helps, but the SleepStagerChambon2018 learns much more quickly and so far reaches higher test performance.

Also I did see a minor improvement in validation performance when combining softmax + CrossEntropyLoss, but so far not as much as what @Tgnassou reported.

More to come soon!

File: examples/plot_sleep_staging_usleep.py
Comment (robintibor) [reply examples/plot_sleep_staging_usleep.py:212]: Did you see this comment here @hubertjb ?

**Comment by hubertjb** [comment]
Trying to conclude on this @robintibor @agramfort @Tgnassou 

1. U-Sleep is much harder to train on the "small-ish" SleepPhysionet. I found some configurations that could somewhat learn more quickly (e.g. model depth=8, weight decay=5e-5, learning rate=1e-5), but they never go much above 50% test balanced accuracy. SleepStagerChambon2018 in contrast can easily reach >70% in under 100 epochs. Using really low learning rates (e.g. 1e-7 like in the paper) does yield very smooth learning curves, but we would have to train models for a lot more than the 100 epochs I tested on to get interesting performance.

2. The balanced sampler seems to help quite a lot. The models learn faster and reach slightly better test performance. Here are some learning curves obtained with SleepStagerChambon2018:

image

3. The self-ensembling also does help, with gains of 5-10% balanced accuracy in my limited experiments on (overfitting) U-Sleep.

Overall, I'm confident the model implementation is accurate. I went through the original keras implementation and made sure the dimensions and the number of parameters were exactly the same (there are now unit tests for these). The model does not learn as well on SleepPhysionet, but my guess is this is an issue with the scale of the data and the training budget rather than the model itself.

**Comment by agramfort** [comment]
Great summary. I suggest to merge. Thanks @hubertjb for finishing this and @Tgnassou and @l-omar-chehab for the ground work

**Comment by Tgnassou** [comment]
Very nice ! Thanks you @hubertjb !

**Comment by agramfort** [comment]
@robintibor the green button is yours !

**Comment by robintibor** [comment]
There were some conflicts now due to other PR merge, which I have resolved. 
More importantly @hubertjb , it seems the comment regarding applying softmax and using crossentropyloss has not been addressed I think (https://github.com/braindecode/braindecode/pull/282/#discussion_r672987582) ? Crossentropyloss will apply softmax again, so it is not correct what is happening there I think... Maybe even `apply_softmax=False` could already solve it? Or do I have some misunderstanding? :)

**Comment by hubertjb** [comment]
Thanks @robintibor for fixing the conflicts and for pointing that out! I forgot to change that line - it should indeed be `apply_softmax=False`.

**Comment by robintibor** [comment]
Great stuff, thanks @Tgnassou @l-omar-chehab @hubertjb , merged now

**Comment by agramfort** [comment]
awesome team effort @Tgnassou @l-omar-chehab @hubertjb 🚀 🚀 🚀 🚀 👏 👏 👏 👏

**Pull request merged**
